plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,36 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+			dnsErrBufOccured := false
+			var perr *dns.Error
+
+			if errors.As(err, &perr) {
+				if errors.Is(err, dns.ErrBuf) {
+					dnsErrBufOccured = true
+				}
+			}
+
+			if proto == "udp" && ((strings.Contains(err.Error(), "overflow")) || dnsErrBufOccured) {
+				newRet := state.Req.Copy()
+
+				newRet.RecursionAvailable = ret.RecursionAvailable
+				newRet.Response = true
+
+				// Set TC bit to indicate truncation.
+				newRet.Truncated = true
+				ret = newRet
+
+				// break here only if response message id matches the request's message id.
+				if state.Req.Id == ret.Id {
+					break
+				}
```

Comment:
This case should be checked before making a copy of the response.  IIUC, it should be inverted and checked earlier.  Something like the following:

```
// drop response if its message id doesn't match the request message id.
if state.Req.Id != ret.Id {
  continue 
}
```

Reply:
Moved this if condition up. Copying the original request state only if the resp message ID matches request message ID.